ISPN-5243 Configuration Value wrappers

tristantarrant opened the pull request:
https://issues.jboss.org/browse/ISPN-5243

- Create "AttributeDefinitions" which describe what configuration attributes look like (name, default value, type, mutable). These defs are "public" so that they can be consumed by others (e.g. Infinispan subsystem in server)
- Attributes are created from definitions and can be either protected (immutable) or unprotected. Also listeners can be attached to react to changes.
- Group Attributes into AttributeSets to manage them together

This feature allows partial configuration overlays and is a stepping stone towards proper template support.

galderz commented:
Looking into this...

galderz commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/Attribute.java:
```diff
@@ -0,0 +1,211 @@
+ * @author Tristan Tarrant
+ * @since 7.2
+ */
+public final class Attribute<T> implements Cloneable {
```
Instead of relying on `Cloneable`, could we use a copy constructor/factory instead? See Bloch's Effective Java, Item 11.

galderz replied:
One benefit is that it'd allow fields to be `final`

tristantarrant replied:
The number of types that we really support within configuration attributes is quite limited: primitive wrappers and Classes (which are immutable) and Properties (and TypedProperties). For all other mutable types the user should really supply an initializer (or factory if you prefer) which constructs the instance for you. It would probably be nicer to drop "Cloneable" and provide a bit of code to handle all known types, defer to an initializer if provided and throw an exception otherwise (so that we won't have any "undefined" behaviour).

tristantarrant replied:
Noticed that the jboss dmr stuff (written by master David Lloyd) uses Cloneable for similar purposes. It makes the code simpler and works for our use case so no need to overcomplicate things.

galderz replied:
As you wish, as long as you've looked at it from all angles, I'm happy with it. With the classes being final, a lot of problems are avoided anyway.

galderz commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/Attribute.java:
```diff
@@ -0,0 +1,211 @@
+      return (Long) value;
+   }
+
+   public <K> K asObject(Class<K> klass) {
```
`klass` unused, do you need it?

wburns replied:
The only benefit would be if you inlined the call by avoiding an explicit cast, otherwise as you say the K can be resolved from the compiler checking the type that the variable is defined as.

tristantarrant replied:
noted.

galderz commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSet.java:
```diff
@@ -0,0 +1,196 @@
+      return attributes.containsKey(name);
+   }
+
+   public <T extends Object> boolean contains(AttributeDefinition<T> def) {
```
I don't think `<T extends Object>` is needed, `<T>` is enough.

tristantarrant replied:
Again some leftover from experiments

galderz commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSet.java:
```diff
@@ -0,0 +1,196 @@
+   public AttributeSet(String name, AttributeSet attributeSet, AttributeDefinition<?>[] attributeDefinitions) {
+      this.name = name;
+      this.attributes = new LinkedHashMap<>(attributeDefinitions.length + attributeSet.attributes.size());
+      for (Attribute<? extends Object> attribute : attributeSet.attributes.values()) {
```
Same thing here, `?` is enough.

Time to use a modern IDE which shows you this right in your screen? :tongue: :wink:

tristantarrant replied:
Again, that stuff is some leftover from trying to get the compiler to cast generic classes properly without resorting to the ugly "double cast" (e.g. (Class<X>)(Class<?>)MyClass.class)

galderz commented on commons/src/main/java/org/infinispan/commons/marshall/AbstractMarshaller.java:
```diff
@@ -1,5 +1,6 @@
 package org.infinispan.commons.marshall;
 
+import org.infinispan.commons.CacheException;
```
Import not needed

galderz commented on core/src/main/java/org/infinispan/configuration/cache/AsyncConfiguration.java:
```diff
@@ -11,19 +14,32 @@
+   private final Attribute<Integer> replicationQueueMaxElements;
+   private final Attribute<Boolean> useReplicationQueue;
+
+   private AttributeSet attributes;
```
Why do you want to keep this attribute set around?

tristantarrant replied:
It is everywhere, it is used for simpler equals, hashCode and toString methods

galderz replied:
Do we really need it? It's extra memory consumption by these objects and duplication. Plus, there's a risk someone might add a attribute and forget to add it to the attribute set. I'm not convinced of its necessity...

tristantarrant replied:
Oh, sorry: we also need it when a *Builder reads() existing configuration. So we can't really do away with it.

galderz commented on core/src/main/java/org/infinispan/configuration/cache/AsyncConfigurationBuilder.java:
```diff
@@ -33,12 +39,12 @@ protected AsyncConfigurationBuilder(ClusteringConfigurationBuilder builder) {
 
    public AsyncConfigurationBuilder asyncMarshalling(boolean async) {
-      this.asyncMarshalling = async;
+      attributes.attribute(MARSHALLING) .set(async);
```
Space formatting

wburns commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:
```diff
@@ -0,0 +1,124 @@
+      return initializer;
+   }
+
+   public void setInitializer(AttributeInitializer<T> initializer) {
```
It seems this is never invoked?  If so we might as well remove it and make the variable final instead.

wburns commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:
```diff
@@ -0,0 +1,124 @@
+public final class AttributeDefinition<T> implements Cloneable {
+   private final String name;
+   private final T defaultValue;
+   private boolean immutable;
```
This can be final.

galderz commented on core/src/main/java/org/infinispan/configuration/cache/Configuration.java:
```diff
@@ -150,27 +150,18 @@ public CompatibilityModeConfiguration compatibility() {
-            ", sites=" + sitesConfiguration +
-            ", compatibility=" + compatibilityConfiguration +
-            '}';
+      return "Configuration [clusteringConfiguration=" + clusteringConfiguration + ", customInterceptorsConfiguration="
```
Is this just a formatting change? :)

The PR is already big enough to add cosmetic changes into it :)

galderz commented on core/src/main/java/org/infinispan/configuration/cache/Configuration.java:
```diff
@@ -201,7 +192,7 @@ public int hashCode() {
       return result;
    }
-   
+
```
Same thing...

tristantarrant replied:
Ahem, I'm removing some spaces

wburns commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:
```diff
@@ -0,0 +1,124 @@
+   private final AttributeValidator<T> validator;
+   private final Class<?> type;
+
+   AttributeDefinition(String name, T initialValue, Class<T> type, boolean immutable, AttributeValidator<T> validator, AttributeInitializer<T> initializer) {
```
To allow for maximum flexibility

1. The validator should be defined as AttributeValidator<? super T>
2. The initializer should be defined as AttributeInitializer<? extends T>

galderz replied:
+1, well spotted.

galderz commented on core/src/main/java/org/infinispan/util/logging/Log.java:
```diff
@@ -1,5 +1,33 @@
 package org.infinispan.util.logging;
 
+import static org.jboss.logging.Logger.Level.DEBUG;
```
ROFL! You put the static imports at the top, and @slaskawi  puts them at the bottom in his PRs, can we stop this madness plz? :)

wburns commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:
```diff
@@ -0,0 +1,124 @@
+      private final T defaultValue;
+      private AttributeInitializer<T> initializer;
+      private AttributeValidator<T> validator;
+      private Class<?> type;
```
why Class<?> ?  Also it seems this should be final.

tristantarrant replied:
Ok

wburns commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSet.java:
```diff
@@ -0,0 +1,196 @@
+
+   @Override
+   public void attributeChanged(Attribute<Object> newValue, Attribute<Object> oldValue) {
+      // TODO
```
still need something here it seems ?

wburns replied:
Or this is going to be done at a later point I am guessing then ?

tristantarrant replied:
Yep.

wburns commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSet.java:
```diff
@@ -0,0 +1,196 @@
+
+   public AttributeSet(String name, AttributeSet attributeSet, AttributeDefinition<?>[] attributeDefinitions) {
+      this.name = name;
+      this.attributes = new LinkedHashMap<>(attributeDefinitions.length + attributeSet.attributes.size());
```
Does this need to be a LinkedHashMap ?

tristantarrant replied:
no :)

tristantarrant commented:
Hi guys, repushed. I've added nearly all of your suggestions, and dropped the .asType() method and I'm just using get() now which does the right thing.

galderz commented:
Thanks @tristantarrant, one final comment above...

galderz commented:
Integrated, thanks @tristantarrant
